expertsystem/amplitude/model.py

```diff
@@ -0,0 +1,349 @@
+"""Data objects for an `~expertsystem.amplitude` model."""
+
+
+from abc import ABC
+from collections import abc
+from dataclasses import dataclass, field
+from enum import Enum, auto
+from typing import (
+    Any,
+    Callable,
+    Dict,
+    Iterable,
+    Iterator,
+    List,
+    Optional,
+    Sequence,
+    Set,
+)
+
+from expertsystem.data import Particle, ParticleCollection
+
+
+@dataclass
+class FitParameter(abc.Hashable):
+    name: str
+    value: float = 0.0
+    is_fixed: bool = False
+
+    def __hash__(self) -> int:
+        return hash(str(self))
+
+
+class FitParameters(abc.Mapping):
+    def __init__(
+        self, parameters: Optional[Iterable[Particle]] = None
+    ) -> None:
+        self.__parameters: Dict[str, FitParameter] = dict()
+        if parameters is not None:
+            if not isinstance(parameters, (list, set, tuple)):
+                raise ValueError(
+                    f"Cannot construct a {self.__class__.__name__} "
+                    f"from a {parameters.__class__.__name__}"
+                )
+            self.__parameters.update(
+                {
+                    par.name: par
+                    for par in parameters
+                    if isinstance(par, FitParameter)
+                }
+            )
+
+    @property
+    def parameter_names(self) -> Set[str]:
+        return set(self.__parameters)
+
+    def __getitem__(self, name: str) -> FitParameter:
+        return self.__parameters[name]
+
+    def __iter__(self) -> Iterator[str]:
+        return self.__parameters.__iter__()
+
+    def __len__(self) -> int:
+        return len(self.__parameters)
+
+    def add(self, parameter: FitParameter) -> None:
+        _assert_arg_type(parameter, FitParameter)
+        if parameter.name in self.__parameters:
+            raise KeyError(f'Parameter "{parameter.name}" already exists')
+        self.__parameters[parameter.name] = parameter
+
+    def filter(  # noqa: A003
+        self, function: Callable[[FitParameter], bool]
+    ) -> "FitParameters":
+        """Search by `FitParameter` properties with a :code:`lambda` function."""
+        return FitParameters(
+            {parameter for parameter in self.values() if function(parameter)}
+        )
+
+    def register_parameter(
+        self, name: str, value: float, fix: bool
+    ) -> FitParameter:
+        parameter = FitParameter(name=name, value=value, is_fixed=fix)
+        self.add(parameter)
+        return parameter
```

Comment:
I would remove that functionality as its not needed. So just use `add(FitParameter(...))`

Reply:
Problem is that you usually want to also register the parameter. So then you would always have to do:
```python
# parameters: FitParameters
par = FitParameter(...)
parameters.add(par)
```
So you can also turn it around: why make the `add` method public?

Reply:
Removed `register_parameter` 84eebf79